From 3f5cd0ebcc77ee7fecbce1763e25cd2d8502e9b8 Mon Sep 17 00:00:00 2001
From: Yumi Yukimura <me.cafebabe@gmail.com>
Date: Sun, 4 May 2025 15:23:57 +0800
Subject: [PATCH 03/10] kernel: Enhancements to searching DTBs during DTB image
 generation

* Allow wildcard pattern when not using QCOM merge_dtbs
* Support searching in specified directory
* Add variable that accepts a list of such patterns

Change-Id: I35da8ed19447b4ab69142152d7dc8e60bc55bf14
---
 build/tasks/kernel.mk | 16 +++++++++++-----
 1 file changed, 11 insertions(+), 5 deletions(-)

diff --git a/build/tasks/kernel.mk b/build/tasks/kernel.mk
index 2452f381..aa79c84c 100644
--- a/build/tasks/kernel.mk
+++ b/build/tasks/kernel.mk
@@ -75,7 +75,9 @@
 #                                          kernel because headers should match
 #
 #   TARGET_MERGE_DTBS_WILDCARD         = Optional, limits the .dtb files used to generate the
-#                                          final DTB image when using QCOM's merge_dtbs script.
+#                                          final DTB image. Deprecated.
+#   TARGET_DTB_LIST_WILDCARD           = Optional, similar to the above one, but allows multiple
+#                                          patterns.
 
 ifneq ($(TARGET_NO_KERNEL),true)
 ifneq ($(TARGET_NO_KERNEL_OVERRIDE),true)
@@ -88,7 +90,11 @@ RECOVERY_DEFCONFIG := $(TARGET_KERNEL_RECOVERY_CONFIG)
 VARIANT_DEFCONFIG := $(TARGET_KERNEL_VARIANT_CONFIG)
 SELINUX_DEFCONFIG := $(TARGET_KERNEL_SELINUX_CONFIG)
 # dtb generation - optional
-TARGET_MERGE_DTBS_WILDCARD ?= *
+ifneq ($(TARGET_MERGE_DTBS_WILDCARD),)
+TARGET_DTB_LIST_WILDCARD ?= $(TARGET_MERGE_DTBS_WILDCARD)
+else
+TARGET_DTB_LIST_WILDCARD ?= *
+endif
 # recovery modules.load fallback - optional
 BOARD_RECOVERY_KERNEL_MODULES_LOAD ?= $(BOARD_RECOVERY_RAMDISK_KERNEL_MODULES_LOAD)
 
@@ -667,10 +673,10 @@ ifeq ($(BOARD_USES_QCOM_MERGE_DTBS_SCRIPT),true)
 	$(hide) find $(DTBS_OUT) -type f -name "*.dtb*" | xargs rm -f
 	mv $(DTB_OUT)/arch/$(KERNEL_ARCH)/boot/dts/vendor/*/*.dtb $(DTB_OUT)/arch/$(KERNEL_ARCH)/boot/dts/vendor/*/*.dtbo $(DTBS_BASE)/
 	PATH=$(abspath $(HOST_OUT_EXECUTABLES)):$${PATH} python3 $(BUILD_TOP)/vendor/lineage/build/tools/merge_dtbs.py --base $(DTBS_BASE) --techpack $(DTB_OUT)/arch/$(KERNEL_ARCH)/boot/dts/vendor/qcom --out $(DTBS_OUT)
-	cat $(shell find $(DTBS_OUT) -type f -name "${TARGET_MERGE_DTBS_WILDCARD}.dtb" | sort) > $@
-else
-	cat $(shell find $(DTB_OUT)/arch/$(KERNEL_ARCH)/boot/dts -type f -name "*.dtb" | sort) > $@
 endif # BOARD_USES_QCOM_MERGE_DTBS_SCRIPT
+	@rm -f $@
+	$(foreach dtb,$(TARGET_DTB_LIST_WILDCARD),\
+		cat `find $(DTB_OUT)/arch/$(KERNEL_ARCH)/boot/dts/$(dir $(dtb)) -type f -name "$(notdir $(dtb)).dtb" | sort` >> $@;)
 endif # BOARD_DTB_CFG
 	$(hide) touch -c $(DTB_OUT)
 endif # !TARGET_WANTS_EMPTY_DTB
-- 
2.39.5

